libraries/AP_InertialSensor/AP_InertialSensor_MPU6000.cpp

```diff
@@ -158,6 +166,193 @@ extern const AP_HAL::HAL& hal;
 #define MPU6000_REV_D8                          0x58    // 0101			1000
 #define MPU6000_REV_D9                          0x59    // 0101			1001
 
+#define MPU6000_SAMPLE_SIZE 12
+#define MPU6000_MAX_FIFO_SAMPLES 3
+
+#define int16_val(v, idx) ((int16_t)(((uint16_t)v[2*idx] << 8) | v[2*idx+1]))
+#define uint16_val(v, idx)(((uint16_t)v[2*idx] << 8) | v[2*idx+1])
+
+/* SPI bus driver implementation */
+void AP_MPU6000_BusDriver_SPI::init(bool &fifo_mode, uint8_t &max_samples)
+{
+    fifo_mode = false;
+    _error_count = 0;
+    _spi = hal.spi->device(AP_HAL::SPIDevice_MPU6000);
+    // Disable I2C bus if SPI selected (Recommended in Datasheet
+    write8(MPUREG_USER_CTRL, BIT_USER_CTRL_I2C_IF_DIS);
+    /* maximum number of samples read by a burst
+     * a sample is an array containing :
+     * gyro_x
+     * gyro_y
+     * gyro_z
+     * accel_x
+     * accel_y
+     * accel_z
+     */
+    max_samples = 1;
+};
+
+void AP_MPU6000_BusDriver_SPI::read8(uint8_t reg, uint8_t *val)
+{
+    uint8_t tx[2];
+    uint8_t rx[2];
+
+    tx[0] = reg;
+    tx[1] = 0;
+    _spi->transaction(tx, rx, 2);
+
+    *val = rx[1];
+}
+
+void AP_MPU6000_BusDriver_SPI::write8(uint8_t reg, uint8_t val)
+{
+    uint8_t tx[2];
+    uint8_t rx[2];
+
+    tx[0] = reg;
+    tx[1] = val;
+    _spi->transaction(tx, rx, 2);
+}
+
+void AP_MPU6000_BusDriver_SPI::set_bus_speed(AP_HAL::SPIDeviceDriver::bus_speed speed)
+{
+    _spi->set_bus_speed(speed);
+}
+
+void AP_MPU6000_BusDriver_SPI::read_burst(uint8_t *samples,
+                                         AP_HAL::DigitalSource *_drdy_pin,
+                                         uint8_t &n_samples)
+{
+    /* one resister address followed by seven 2-byte registers */
+    struct PACKED {
+        uint8_t cmd;
+        uint8_t int_status;
+        uint8_t d[14];
+    } rx, tx = { cmd : MPUREG_INT_STATUS | 0x80, };
+
+    _spi->transaction((const uint8_t *)&tx, (uint8_t *)&rx, sizeof(rx));
+
+    /*
+      detect a bad SPI bus transaction by looking for all 14 bytes
+      zero, or the wrong INT_STATUS register value. This is used to
+      detect a too high SPI bus speed.
+    */
+    uint8_t i;
+    for (i=0; i<14; i++) {
+        if (rx.d[i] != 0) break;
+    }
+    if ((rx.int_status&~0x6) != (_drdy_pin==NULL?0:BIT_RAW_RDY_INT) || i == 14) {
+        // likely a bad bus transaction
+        if (++_error_count > 4) {
+            set_bus_speed(AP_HAL::SPIDeviceDriver::SPI_SPEED_LOW);
```

Comment:
huumn.... this will get lost when you read another device. It should be a property of ***this*** bus for ***this*** device, not the spi manager.